Fix and improve Makefile / CMake

SamantazFox opened the pull request:
### Requirements

Building Marlin without the Arduino IDE

### Description

Multiple fixes on the Makefile, so Marlin can be properly built without the Arduino IDE
(i.e, only with the core(s) and the source files)

### Benefits

* Marlin can be properly built without the Arduino IDE nor PlatformIO
* Some common Makefile bugs are fixed

### Related Issues

None found


NB: same as PR #19633, but on the right branch this time!

SamantazFox committed:
Makefile: Change the default board id to match Configuration.h

change the default board to BOARD_RAMPS_14_EFB (1020) in the Makefile to avoid
the following error message on a clean build (no variables given to make):

src/pins/pins.h:743:6: error: \#error "Unknown MOTHERBOARD value set in Configuration.h"
     \#error "Unknown MOTHERBOARD value set in Configuration.h"
      ^

SamantazFox committed:
Makefile: disable U8GLIB and TMC libraries

Disable these two libraries because they're not provided with Arduino nor Marlin
by default, so a no-arguments `make` simply fails.

SamantazFox committed:
Makefile: Fix some libraries paths added to VPATH

SamantazFox committed:
Makefile: Add Tone.cpp to the CXX source files

Tone.cpp (from ADRUINO_DIR/hardware/arduino/avr/cores/arduino/) is required when
the speaker function is enabled. Otherwise, the linker complains:

applet/src/libs/buzzer.o: In function `Buzzer::tick()':
buzzer.cpp:(.text._ZN6Buzzer4tickEv+0xb2): undefined reference to
	`tone(unsigned char, unsigned int, unsigned long)'
collect2: error: ld returned 1 exit status

SamantazFox committed:
Makefile: formating, wrap long lines when possible

SamantazFox committed:
Makefile: Use CXX to build .elf target

Most of the time, CC resolves to avr-gcc, which is the C compiler, and hence
won't be able to properly link C++ objects (It won't pass the proper flags to
the linker). So use CXX instead.

SamantazFox committed:
Makefile: Use PROG_MCU instead of MCU for AVRdude

AVRdude and GCC's MCU names are not the same, so we need to define a different
variable for AVRdude's part name.

All mappings were added for existing boards.

thinkyhead commented:
As a work-in-progress, what is left to do before this is ready? Does it need some additional testing?

SamantazFox commented:
I recall an issue with the EEPROM library, which I'm trying to reproduce. After that, it'll be done.

As for tests, yes, that'd be cool to have some feedback on the changes, across different
hardware configs because I can't test them all by myself. It would be a nice addition to make those tests part of the auto PR checks (currently, they only do PlatformIO validation).

SamantazFox committed:
Makefile: Define F_CPU as unsigned long

Respect avr-libc's manapage regarding "F_CPU" (see `man 3 F_CPU`): this number
is defined as an 'unsigned long' (UL suffix).

In order to avoid issues, the UL suffix is added in one place:
the CDEFS declaration.

SamantazFox commented:
Mmmh, can't reproduce the EEPROM thing.

Not related to eeprom, I'm wondering if I should add something to automatically
craft the architecture-specific libraries' path?

As you can see [here](https://github.com/SamantazFox/Marlin/blob/makefile-patches/Marlin/Makefile#L604) and [here](https://github.com/SamantazFox/Marlin/blob/makefile-patches/Marlin/Makefile#L619), the libraries' path are hard-coded, and I doubt they'll work with sam, samd or curie-based platforms.

thinkyhead committed:
Fix up CMakeLists.txt too

thinkyhead commented:
Under macOS this is what I currently get (after setting up `ARDUINO_INSTALL_DIR` and `ARDUINO_USER_DIR` paths):
```
make: *** No rule to make target `applet/arduino/wiring.o', needed by `applet/Marlin.elf'.  Stop.
```
I'm not sure where to begin to get past this, but I'll try some Googling.

SamantazFox commented:
> Under macOS this is what I currently get (after setting up `ARDUINO_INSTALL_DIR` and `ARDUINO_USER_DIR` paths):
> 
> ```
> make: *** No rule to make target `applet/arduino/wiring.o', needed by `applet/Marlin.elf'.  Stop.
> ```
> 
> I'm not sure where to begin to get past this, but I'll try some Googling.

This means that you're not able to compile the file `wiring.c`, which should be under `${ARDUINO_INSTALL_DIR}/hardware/arduino/avr/cores/arduino/`

SamantazFox commented on buildroot/share/cmake/CMakeLists.txt:
```diff
@@ -97,7 +97,7 @@ setup_motherboard(${PROJECT_NAME} ${CMAKE_CURRENT_LIST_DIR}/../../../Marlin)
 #====================================================================#
 
 file(GLOB SOURCES "../../../src/*.cpp")
```
Shoudln't that one be `../../../Marlin` instead of `../../../src`, too?

thinkyhead replied:
🤷 I haven't gotten far enough to check on that. And it seems like this `GLOB` value isn't going to be relevant to our new complex file layout.

SamantazFox replied:
Maybe `../../../Marlin/**/*.cpp` then?

thinkyhead replied:
If that style of path can apply in this context….

SamantazFox replied:
Okay, so `file(GLOB_RECURSE SOURCES "../../../Marlin/*.cpp")` is what we want here.

See: https://cmake.org/cmake/help/latest/command/file.html#glob-recurse

thinkyhead replied:
Good to know, thanks! Ultimately, it would be nice if these make files could filter sources according to enabled features, the way we currently do with PlatformIO.

SamantazFox replied:
Unless some sort of configure script extracts the Configuration.h defines and feed them to the Makefile, I don't see how this could be done.

thinkyhead committed:
Fix sources pattern

thinkyhead committed:
Assign paths per platform

thinkyhead commented:
The macOS build is working as long as I provide all the paths. To help out others I've added default paths for each platform.

thinkyhead committed:
Comment tweaks

thinkyhead committed:
ibid.

SamantazFox commented:
> The macOS build is working as long as I provide all the paths. To help out others I've added default paths for each platform.

Nice :)

SamantazFox commented:
Btw, is it possible to add CI tests for the Makefile?
The current pipeline only validates the PlatformIO toolchain.

thinkyhead commented:
> Btw, is it possible to add CI tests for the Makefile?

Anything is possible. But what aspect of Makefile do we need to test for *every single commit* that we cannot test through existing methods (other than checking that the Makefile itself is not broken)?

SamantazFox commented:
> > Btw, is it possible to add CI tests for the Makefile?
> 
> Anything is possible. But what aspect of Makefile do we need to test for _every single commit_ that we cannot test through existing methods (other than checking that the Makefile itself is not broken)?

In order to test if we can build Marlin without PlatformIO nor the Arduino IDE?
